beacon_node/store/src/hot_cold_store.rs

```diff
@@ -3100,23 +3438,39 @@ impl StoreItem for HotStateSummary {
 
 impl HotStateSummary {
     /// Construct a new summary of the given state.
-    pub fn new<E: EthSpec>(state_root: &Hash256, state: &BeaconState<E>) -> Result<Self, Error> {
+    pub fn new<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>>(
+        store: &HotColdDB<E, Hot, Cold>,
+        state_root: &Hash256,
+        state: &BeaconState<E>,
+        storage_strategy: StorageStrategy,
+    ) -> Result<Self, Error> {
         // Fill in the state root on the latest block header if necessary (this happens on all
         // slots where there isn't a skip).
         let latest_block_root = state.get_latest_block_root(*state_root);
-        let epoch_boundary_slot = state.slot() / E::slots_per_epoch() * E::slots_per_epoch();
-        let epoch_boundary_state_root = if epoch_boundary_slot == state.slot() {
-            *state_root
+
+        let get_state_root = |slot| {
+            if slot == state.slot() {
+                Ok::<_, Error>(*state_root)
+            } else {
+                Ok(get_ancenstor_state_root(store, state, slot)?.ok_or(
+                    HotColdDBError::HdiffGetPriorStateRootError(state.slot(), slot),
+                )?)
+            }
+        };
+        let diff_base_slot = storage_strategy.diff_base_slot();
+        let diff_base_state_root = if let Some(diff_base_slot) = diff_base_slot {
+            DiffBaseStateRoot::new(diff_base_slot, get_state_root(diff_base_slot)?)
         } else {
-            *state
-                .get_state_root(epoch_boundary_slot)
-                .map_err(HotColdDBError::HotStateSummaryError)?
+            DiffBaseStateRoot::zero()
```

Comment:
Maybe we should store a non-zero value here. We don't want to accidentally try loading a 0x0 state with the backwards iterator (get_ancestor_state_root).

We could even use an `Option<DiffBaseState>` for snapshots? (Need to make it serialize)

Or we could use something like:

```rust
DiffBaseStateRoot {
    slot: u64::MAX,
    state_root: Hash256::repeat_byte(0xff),
}
```

Reply:
This is way better than both actually:

```rust
#[derive(Encode, Decode)]
enum DiffBaseStateRoot {
   Snapshot,
   BaseState {
      slot: Slot,
      state_root: Hash256,
   }
}
```

Reply:
Done with `OptionalDiffBaseState`